src/neptune/main.d

```diff
@@ -0,0 +1,342 @@
+    auto tags = repo.releasedTags().map!(a=>Version(a.name)).array;
+
+    sort(tags);
+
+    cmd("git remote update");
+
+    auto release = autodetect(tags);
+
+    ActionList list;
+
+    if (release.type == release.type.Patch)
+        list = preparePatchRelease(repo, tags, release);
+    else
+    {
+        writefln("This seems to be %s release which is not yet supported, exiting...",
+                 release.type);
+        return;
+    }
+
+    writefln("Actions to be done:");
+
+    foreach (action; list.local_actions)
+        writefln(" * %s (%s)", action.description, action.command);
+
+    writef("\nWould you like to continue? y/n: ");
+
+    if (!readYesNoResponse())
+    {
+        writefln("Aborting...");
+        return;
+    }
+
+
+    writefln("");
+    foreach (action; list.local_actions)
+    {
+        import std.string : strip;
+
+        writefln("%s ...\n\t%s",
+                 action.description,
+                 action.command);
+        writefln("\t%s", strip(cmd(action.command)));
+    }
+
+
+    writefln("Done! Would you like to push the modified references now?");
+    writefln("References to be pushed: %s", list.affected_refs);
+    writef("y/n?: ");
+
+    if (!readYesNoResponse())
+    {
+        writefln("Aborting...");
+        return;
+    }
+
+    foreach (reference; list.affected_refs)
+    {
+        import std.format;
+
+        writefln("Pushing %s ...", reference);
+        cmd(format("git push %s %s", getRemote(), reference));
+    }
+}
+
+
+/*******************************************************************************
+
+    Prepares the actions for a patch release
+
+    Params:
+        repo = repo object
+        tags = array of tags for this repo
+        patch_version = version to release
+
+    Returns:
+        prepared ActionList object with all refs/actions required for this
+        release
+
+*******************************************************************************/
+
+ActionList preparePatchRelease ( ref Repository repo, Version[] tags,
+                                 Version patch_version )
+{
+    import neptune.github;
+
+    import std.format;
+    import std.algorithm;
+    import std.range;
+    import std.stdio;
+
+    auto upstream = getRemote();
+
+    auto list = makeRelease(patch_version, format("%s/%s",
+                                                  upstream,
+                                                  getCurrentBranch()));
+
+    struct Branch
+    {
+        Version ver;
+        string sha;
+
+        alias ver this;
+    }
+
+    bool thisVersion ( Version v )
+    {
+        return v.major == patch_version.major &&
+               v.minor == patch_version.minor;
+    }
+
+    bool newerVersion ( Version v )
+    {
+        return v > patch_version;
+    }
+
+    bool ourMajorRelease ( Version v )
+    {
+        return v.major == patch_version.major &&
+               v.minor.isNull && v.patch.isNull;
+    }
```

Comment:
They are used for the `filter` call which calls them for every single element...